src/plugins/dkim_check.c

```diff
@@ -560,62 +586,71 @@ dkim_module_check (struct dkim_check_result *res)
 				}
 			}
 		}
+	}
 
-		if (cur->res == -1 || cur->key == NULL) {
+	DL_FOREACH (first, cur) {
+		if (cur->ctx == NULL) {
+			continue;
+		}
+		if (cur->res == -1 || (cur->key == NULL && cur->err == NULL && cur->res != DKIM_TRYAGAIN)) {
 			/* Still need a key */
 			all_done = FALSE;
 		}
+		if (cur->res == DKIM_CONTINUE) {
+			got_allow = TRUE;
+		}
 	}
 
 	if (all_done) {
 		DL_FOREACH (first, cur) {
+			const gchar *symbol = NULL;
+			GList *messages = NULL;
+			int symbol_weight = 1;
 			if (cur->ctx == NULL) {
 				continue;
 			}
-
-			if (cur->res == DKIM_CONTINUE) {
-				rspamd_task_insert_result (cur->task,
-						dkim_module_ctx->symbol_allow,
-						cur->mult_allow * 1.0,
-						g_list_prepend (NULL,
-								rspamd_mempool_strdup (cur->task->task_pool,
-										rspamd_dkim_get_domain (cur->ctx))));
-				got_allow = TRUE;
-				sel = NULL;
+			if (!got_allow && cur->res == DKIM_REJECT) {
+				symbol = dkim_module_ctx->symbol_reject;
+				symbol_weight = cur->mult_deny * 1.0;
+				messages = g_list_prepend (messages,
+						rspamd_mempool_strdup (cur->task->task_pool,
+						rspamd_dkim_get_domain (cur->ctx)));
 			}
-			else if (!got_allow) {
-				if (sel == NULL) {
-					sel = cur;
-				}
-				else if (sel->res == DKIM_TRYAGAIN && cur->res != DKIM_TRYAGAIN) {
-					sel = cur;
-				}
+			else if (cur->res == DKIM_CONTINUE) {
+				symbol = dkim_module_ctx->symbol_allow;
+				symbol_weight = cur->mult_allow * 1.0;
+				messages = g_list_prepend (messages,
+						rspamd_mempool_strdup (cur->task->task_pool,
+						rspamd_dkim_get_domain (cur->ctx)));
+			}
+			else if (!got_allow && cur->res == DKIM_PERM_ERROR && cur->err->message != NULL) {
```

Comment:
It should never be NULL in case of DKIM_PERM_ERROR but it doesn't hurt to check this properly.